internal/http_client.go

```diff
@@ -188,3 +280,94 @@ func WithQueryParams(qp map[string]string) HTTPOption {
 		r.URL.RawQuery = q.Encode()
 	}
 }
+
+// RetryConfig specifies how the HTTPClient should retry failing HTTP requests.
+//
+// A request is never retried more than MaxRetries times. If CheckForRetry is nil, all network
+// errors, and all 400+ HTTP status codes are retried. If an HTTP error response contains the
+// Retry-After header, it is always respected. Otherwise retries are delayed with exponential
+// backoff. Set ExpBackoffFactor to 0 to disable exponential backoff, and retry immediately
+// after each error.
+//
+// If MaxDelay is set, retries delay gets capped by that value. If the Retry-After header
+// requires a longer delay than MaxDelay, retries are not attempted.
+type RetryConfig struct {
+	MaxRetries       int
+	CheckForRetry    RetryCondition
+	ExpBackoffFactor float64
+	MaxDelay         *time.Duration
+}
+
+// RetryCondition determines if an HTTP request should be retried depending on its last outcome.
+type RetryCondition func(resp *http.Response, networkErr error) bool
+
+func (rc *RetryConfig) retryDelay(retries int, resp *http.Response, err error) (time.Duration, bool) {
+	if !rc.retryEligible(retries, resp, err) {
+		return 0, false
+	}
+	estimatedDelay := rc.estimateDelayBeforeNextRetry(retries)
+	serverRecommendedDelay := parseRetryAfterHeader(resp)
+	if serverRecommendedDelay > estimatedDelay {
+		estimatedDelay = serverRecommendedDelay
+	}
+	if rc.MaxDelay != nil && estimatedDelay > *rc.MaxDelay {
+		return 0, false
+	}
+	return estimatedDelay, true
+}
+
+func (rc *RetryConfig) retryEligible(retries int, resp *http.Response, err error) bool {
+	if retries >= rc.MaxRetries {
+		return false
+	}
+	if rc.CheckForRetry == nil {
+		return err != nil || resp.StatusCode >= 400
+	}
+	return rc.CheckForRetry(resp, err)
+}
+
+func (rc *RetryConfig) estimateDelayBeforeNextRetry(retries int) time.Duration {
+	if retries == 0 {
+		return 0
+	}
+	delayInSeconds := int64(math.Pow(2, float64(retries)) * rc.ExpBackoffFactor)
+	estimatedDelay := time.Duration(delayInSeconds) * time.Second
+	if rc.MaxDelay != nil && estimatedDelay > *rc.MaxDelay {
+		estimatedDelay = *rc.MaxDelay
+	}
+	return estimatedDelay
+}
+
+var retryTimeClock Clock = &SystemClock{}
+
+func parseRetryAfterHeader(resp *http.Response) time.Duration {
+	if resp == nil {
+		return 0
+	}
+	retryAfterHeader := resp.Header.Get("retry-after")
+	if retryAfterHeader == "" {
+		return 0
+	}
+	delayInSeconds, err := strconv.ParseInt(retryAfterHeader, 10, 64)
```

Comment:
If this returns an error and parsing as a time returns an error, we still use the value of `delayInSeconds`?

It seems like this should be:

	delayInSeconds, err := strconv.ParseInt(retryAfterHeader, 10, 64)
	if err == nil {
		return time.Duration(delayInSeconds) * time.Second
	}
	timestamp, err := http.ParseTime(retryAfterHeader)
	if err == nil {
		return timestamp.Sub(retryTimeClock.Now())
	}
	return 0

Reply:
Good point. I was assuming `ParseInt` returns 0 when there's an error. But looks like that's not always the case.